### ISSUE TITLE
Support refreshing access tokens

### ISSUE THREAD

**Issue by seanh** [ckan/ckanext-oauth2waad#1]
See http://msdn.microsoft.com/en-us/library/azure/dn645542.aspx. Currently ckanext-oauth2waad doesn't attempt to refresh access tokens, or in any way handle expiring access tokens.

Proposed spec:
1. When we receive the access token response from WAAD, we'll store the refresh token and its expiry time
2. On each page load (so only when the user clicks on a link on the site) we'll check the expiry time, if the expiry time is coming up in the next 5 mins or has already passed we'll try to use the refresh token to get a new access token.
3. It's still possible that the user might stay on one page for longer than 5 minutes and have their access token expire without the server having a chance to refresh it on a page load. So whenever CKAN makes an API request to "the resource" (e.g. to request the creation of a dataset), if that request fails for any reason we'll try refreshing the access token and trying the request again.
4. We will also store the **authorization code** that we used to get the access and refresh tokens. Whenever we try to refresh the access token, if the refresh request fails for any reason we'll trying using the authorization code to get a new access token instead.
5. At any point, if both trying to get a new access token using the refresh token and trying to get one using the authorization code fail, we will log the user out of CKAN. The user experience will be: they click on a link or button to go to another CKAN page, instead they are logged out and redirected to the CKAN login page. They can now try to login using WAAD again.
6. If the link/button the user clicked on before they were logged-out was a form submission button, they may lose their work in progress. We could handle this in CKAN by storing "draft" datasets and other objects in the CKAN db. But I suggest this is an edge case and not worth worrying about, at least not at first.

In 3 we do the refresh attempt if the request fails _for any reason_ because the API doesn't seem to give back specific reasons when it denies requests.

2 may slow down CKAN page load times noticeably (the CKAN server has to make an HTTP request to refresh the access token, before it can get on with rendering the page), but this won't happen often, only when the access token expires, which looks like once an hour from what I've seen (though I don't think there's any guarantee that expiry times won't change).

2 could be improved by using JavaScript to do the token refreshing, instead of doing it server-side on page loads. But I'd like to get a non-JavaScript version working first and then progressively enhance if necessary.

Questions:
- Is there an easy way to test/develop this? Since it looks like access tokens last for 1 hour, I will have to wait 1 hour each time I want to test to see what happens. At some point I'll be able to mock the WAAD server, but I can't do that until I've seen how it behaves in each different circumstance.

**Comment by adamamyl** [comment]
In theory, the Access tokens should last for one hour. In testing, I believe that's not always the case. AIUI, this time limit is not (presently) changeable in WAAD at all.

**Comment by adamamyl** [comment]
> Is there an easy way to test/develop this? Since it looks like access tokens last for 1 hour, I will have to wait 1 hour each time I want to test to see what happens. At some point I'll be able to mock the WAAD server, but I can't do that until I've seen how it behaves in each different circumstance.

I can't think of a way, but maybe @stevenlivz knows of an approach?

**Comment by stevenlivz** [comment]
Looks good.  Thoughts - in step 3, if the request gets a 403 it probably needs a token refresh - other than that it would be for another reason that shouldn't require a token refresh (at least initially).

I asked about the Access Token. This expiry time can't be changed :-( Bit of a pain.

**Comment by seanh** [comment]
I've implemented 1 and 2, will work on the others next

**Comment by seanh** [comment]
This is done now (and with lots of tests), what I implemented was:
1. On every page load, check if the access token is going to expire in the next five minutes (or has already expired)
2. If so, attempt to get a new access token using the refresh token.
3. If refreshing the access token fails, we flash an error message to the user warning them that some functionality may not work and suggesting they try logging out and logging back in again. I decided not to forcefully log people out.

I still need to factor out the code for refreshing the access token into a public function that other plugins can call, e.g. to try to refresh the token if an API request fails. So leaving this issue open for now.
